File: LeetCode/219ContainsDuplicateII/ContainsDuplicateII.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # 2 passes -> Time:O(n), Space:O(n)
        d = {}
        for i in range(len(nums)):
            if nums[i] not in d:
                d[nums[i]] = i
            elif abs(d[nums[i]] - i) <= k:
                return True
            else:
                d[nums[i]] = i
        return False

        # # 2 passes -> Time:O(n), Space:O(n)
        # s = set()
        # l, r = 0, 0
        # while r < len(nums):
        #     while r - l <= k and r < len(nums):
        #         if nums[r] in s:
        #             return True
        #         s.add(nums[r])
        #         r += 1
        #     if r == len(nums):
        #         return False
        #     s.remove(nums[l])
        #     l += 1
        # return False

File: LeetCode/219ContainsDuplicateII/ContainsDuplicateII.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        // 2 passes -> Time:O(n), Space:O(n)
        set<int> s;
        int l = 0, r = 0;
        while (r < nums.size()) {
            while (r - l <= k && r < nums.size()) {
                if (s.find(nums[r]) != s.end()) {
                    return true;
                }
                s.insert(nums[r]);
                r++;
            }
            if (r == nums.size()) {
                return false;
            }
            s.erase(nums[l]);
            l++;
        }
        return false;
    }
};
